src/klein/app.py

```diff
@@ -186,9 +194,37 @@ def index(request):
 
         @returns: decorated handler function.
         """
-        segment_count = url.count('/')
-        if url.endswith('/'):
-            segment_count -= 1
+
+        def _urlToString(_url):
+            return _url.asText() if URL is not False else _url
+
+        if URL is not False:
+
+            if isinstance(url, string_types + (binary_type, )):
+                url = URL.fromText(url)
+            if url.absolute or url.port or url.userinfo:
+                warnings.warn(
+                    "Mind to use base URI for an application routing, "
+                    "instead of "
+                    + url.asText(includeSecrets=False))
+                url = url.replace(
+                    host=None,
+                    scheme=None,
+                    port=None,
+                    userinfo=u"",
+                )
+
+            if not url.rooted:
+                url = url.replace(rooted=True)
```

Comment:
actually `werkzeug` raises `ValueError` if no leading slash is present in  url  - see https://github.com/pallets/werkzeug/blob/master/werkzeug/routing.py#L603

my original intent was to avoid forcing users to initialize every `URL` instance, passed to  `Klein.route()`, with `rooted=True` constructor parameter (since `werkzeug` requires it), thus reducing keystrokes; I still think it worth the while, so now making the behavior for plain URLs consistent.

as for the checks for non-path components above - agree that it is better to reproduce original `werkzeug` behavior here, removing them.